libraries/opensk/src/api/persist.rs

```diff
@@ -0,0 +1,367 @@
+        let pin_properties = match self.find(keys::PIN_PROPERTIES)? {
+            None => return Ok(None),
+            Some(pin_properties) => pin_properties,
+        };
+        if pin_properties.len() != 1 + PIN_AUTH_LENGTH {
+            return Err(Ctap2StatusCode::CTAP2_ERR_VENDOR_INTERNAL_ERROR);
+        }
+        Ok(Some(*array_ref![pin_properties, 1, PIN_AUTH_LENGTH]))
+    }
+
+    /// Returns the length of the currently set PIN if defined.
+    #[cfg(feature = "config_command")]
+    fn pin_code_point_length(&self) -> CtapResult<Option<u8>> {
+        let pin_properties = match self.find(keys::PIN_PROPERTIES)? {
+            None => return Ok(None),
+            Some(pin_properties) => pin_properties,
+        };
+        if pin_properties.is_empty() {
+            return Err(Ctap2StatusCode::CTAP2_ERR_VENDOR_INTERNAL_ERROR);
+        }
+        Ok(Some(pin_properties[0]))
+    }
+
+    /// Sets the PIN hash and length.
+    ///
+    /// If it was already defined, it is updated.
+    fn set_pin(
+        &mut self,
+        pin_hash: &[u8; PIN_AUTH_LENGTH],
+        pin_code_point_length: u8,
+    ) -> CtapResult<()> {
+        let mut pin_properties = [0; 1 + PIN_AUTH_LENGTH];
+        pin_properties[0] = pin_code_point_length;
+        pin_properties[1..].clone_from_slice(pin_hash);
+        self.insert(keys::PIN_PROPERTIES, &pin_properties[..])?;
+        // If power fails between these 2 transactions, PIN has to be set again.
+        self.remove(keys::FORCE_PIN_CHANGE)
+    }
+
+    /// Returns the number of failed PIN attempts.
+    fn pin_fails(&self) -> CtapResult<u8> {
+        match self.find(keys::PIN_RETRIES)? {
+            None => Ok(0),
+            Some(value) if value.len() == 1 => Ok(value[0]),
+            _ => Err(Ctap2StatusCode::CTAP2_ERR_VENDOR_INTERNAL_ERROR),
+        }
+    }
+
+    /// Decrements the number of remaining PIN retries.
+    fn incr_pin_fails(&mut self) -> CtapResult<()> {
+        let old_value = self.pin_fails()?;
+        let new_value = old_value.saturating_add(1);
+        self.insert(keys::PIN_RETRIES, &[new_value])
+    }
+
+    /// Resets the number of remaining PIN retries.
+    fn reset_pin_retries(&mut self) -> CtapResult<()> {
+        self.remove(keys::PIN_RETRIES)
+    }
+
+    /// Returns the minimum PIN length, if stored.
+    fn min_pin_length(&self) -> CtapResult<Option<u8>> {
+        match self.find(keys::MIN_PIN_LENGTH)? {
+            None => Ok(None),
+            Some(value) if value.len() == 1 => Ok(Some(value[0])),
+            _ => Err(Ctap2StatusCode::CTAP2_ERR_VENDOR_INTERNAL_ERROR),
+        }
+    }
+
+    /// Sets the minimum PIN length.
+    #[cfg(feature = "config_command")]
+    fn set_min_pin_length(&mut self, min_pin_length: u8) -> CtapResult<()> {
+        self.insert(keys::MIN_PIN_LENGTH, &[min_pin_length])
+    }
+
+    /// Returns the list of RP IDs that may read the minimum PIN length.
```

Comment:
nit: let's specify that it defaults to the empty vector (instead of returning `Option<Vec<u8>>`).

Reply:
Done